### PR TITLE
samples: lvgl: sensor: Add LVGL sample for real-time sensor data chart

### PR DESCRIPTION
**Pull request by kartben**
This adds a new sample for LVGL that displays a real-time chart for accelerometer data acquired from &accel0 sensor.
This adds the appropriate overlays so that the sample can also run on native POSIX and native sim using emulated BMI160.

It's meant to show how to generically fetch sensor data and display it on a GUI. It also shows how to piggy back the LVGL built-in timer functionality.

📃 LIVE DOC: https://builds.zephyrproject.io/zephyr/pr/63375/docs/samples/modules/lvgl/acceleration_chart/README.html

### CODE REVIEW

File: samples/modules/lvgl/acceleration_chart/src/main.c [hunk 1]
```diff
@@ -0,0 +1,96 @@
...
+#include <zephyr/drivers/sensor.h>
+
+#include <lvgl.h>
+#include <stdio.h>
+#include <string.h>
+#include <zephyr/kernel.h>
+
+#define LOG_LEVEL CONFIG_LOG_DEFAULT_LEVEL
+#include <zephyr/logging/log.h>
+LOG_MODULE_REGISTER(app);
```
Comment (faxe1008) [inline samples/modules/lvgl/acceleration_chart/src/main.c:19]: `CONFIG_LOG_DEFAULT_LEVEL` can be added in here directly.

File: samples/modules/lvgl/acceleration_chart/src/main.c [hunk 2]
```diff
@@ -0,0 +1,96 @@
...
+
+	lv_chart_set_point_count(chart1, 50);
+
+	/* Do not display point markers on the data */
+	lv_obj_set_style_size(chart1, 0, LV_PART_INDICATOR);
+}
+
+int main(void)
+{
+	display_dev = DEVICE_DT_GET(DT_CHOSEN(zephyr_display));
```
Comment (faxe1008) [inline samples/modules/lvgl/acceleration_chart/src/main.c:74]: `display_dev` is not accessed anywhere else, maybe move it into the function.

File: samples/modules/lvgl/acceleration_chart/src/main.c [hunk 3]
```diff
@@ -0,0 +1,96 @@
...
+	if (!device_is_ready(display_dev)) {
+		LOG_ERR("Device not ready, aborting test");
+		return 0;
```
Comment (faxe1008) [inline samples/modules/lvgl/acceleration_chart/src/main.c:77]: Had that mistake in the other demo one also, I think returning `-ENODEV` would be more fitting.

File: samples/modules/lvgl/acceleration_chart/src/main.c [hunk 4]
```diff
@@ -0,0 +1,96 @@
...
+	}
+
+	sensor = DEVICE_DT_GET(ACCEL0_NODE);
+	if (!device_is_ready(sensor)) {
+		LOG_ERR("Device %s is not ready\n", sensor->name);
+		return 0;
```
Comment (faxe1008) [inline samples/modules/lvgl/acceleration_chart/src/main.c:83]: Same in here.

File: samples/modules/lvgl/acceleration_chart/src/main.c [hunk 5]
```diff
@@ -0,0 +1,96 @@
...
+	lv_chart_set_update_mode(chart1, LV_CHART_UPDATE_MODE_SHIFT);
+
+	ser_x = lv_chart_add_series(chart1, lv_palette_main(LV_PALETTE_RED),
+				    LV_CHART_AXIS_PRIMARY_Y);
+	ser_y = lv_chart_add_series(chart1, lv_palette_main(LV_PALETTE_BLUE),
+				    LV_CHART_AXIS_PRIMARY_Y);
+	ser_z = lv_chart_add_series(chart1, lv_palette_main(LV_PALETTE_GREEN),
+				    LV_CHART_AXIS_PRIMARY_Y);
+
+	lv_chart_set_point_count(chart1, 50);
```
Comment (faxe1008) [inline samples/modules/lvgl/acceleration_chart/src/main.c:66]: If you want to be fancy, for bigger/smaller screens give the option to configure via Kconfig?

File: samples/modules/lvgl/acceleration_chart/src/main.c [hunk 6]
```diff
@@ -0,0 +1,96 @@
...
+	if (rc < 0) {
+		LOG_ERR("ERROR: Update failed: %d\n", rc);
+	} else {
+		lv_chart_set_next_value(chart1, ser_x, sensor_value_to_double(&accel[0]));
+		lv_chart_set_next_value(chart1, ser_y, sensor_value_to_double(&accel[1]));
+		lv_chart_set_next_value(chart1, ser_z, sensor_value_to_double(&accel[2]));
+	}
+}
+
+void accelerometer_chart(lv_obj_t *parent)
```
Comment (faxe1008) [inline samples/modules/lvgl/acceleration_chart/src/main.c:50]: Nit: personal prefrence but I think there is a verb missing in the function name, maybe `init` or  `create`.

File: samples/modules/lvgl/acceleration_chart/src/main.c [hunk 7]
```diff
@@ -0,0 +1,96 @@
...
+
+#define LOG_LEVEL CONFIG_LOG_DEFAULT_LEVEL
+#include <zephyr/logging/log.h>
+LOG_MODULE_REGISTER(app);
+
+static lv_obj_t *chart1;
+static lv_chart_series_t *ser_x;
+static lv_chart_series_t *ser_y;
+static lv_chart_series_t *ser_z;
+lv_timer_t *fetch_and_display_timer;
```
Comment (faxe1008) [inline samples/modules/lvgl/acceleration_chart/src/main.c:25]: Is there any reason you did not make this one static?

**Review by faxe1008 — CHANGES_REQUESTED** [review]
Mostly nits, good stuff. 
Also thanks for mentioning the preview of the docs. Did not know about that one. Neat.

File: samples/modules/lvgl/acceleration_chart/src/main.c (hunk 6 above)
Comment (kartben) [reply samples/modules/lvgl/acceleration_chart/src/main.c:50]: oh wow yeah, totally agreed - so much for hastily cleaning up a very old PoC into a sample :)

File: samples/modules/lvgl/acceleration_chart/src/main.c (hunk 7 above)
Comment (kartben) [reply samples/modules/lvgl/acceleration_chart/src/main.c:25]: +1

File: samples/modules/lvgl/acceleration_chart/src/main.c (hunk 5 above)
Comment (kartben) [reply samples/modules/lvgl/acceleration_chart/src/main.c:66]: mmh yeah I like the idea of showing configurability. Maybe I'll make sensor sampling rate an option as well.

File: samples/modules/lvgl/acceleration_chart/src/main.c (hunk 3 above)
Comment (kartben) [reply samples/modules/lvgl/acceleration_chart/src/main.c:77]: ahah yes I took some of your code and mistakes too, I guess. should have remembered about @fabiobaltieri's late comments on your PR.

File: samples/modules/lvgl/acceleration_chart/src/main.c (hunk 6 above)
Comment (kartben) [reply samples/modules/lvgl/acceleration_chart/src/main.c]: i went for `create_...` as this seems to be what's used for the demos.

File: samples/modules/lvgl/acceleration_chart/src/main.c (hunk 7 above)
Comment (kartben) [reply samples/modules/lvgl/acceleration_chart/src/main.c]: done and renamed to `sensor_timer`, also renamed the callback to `sensor_timer_cb` (again, I went ffor the convention used in the demos)

**Comment by kartben** [comment]
@faxe1008 Hopefully I didn't make the diff unecessarily hard to check compared to previous revision: I ended up renaming the sample from acceleration_chart to accelerometer_chart.
added `CONFIG_SAMPLE_ACCEL_SAMPLING_RATE` and `CONFIG_SAMPLE_CHART_POINTS_PER_SERIES` Kconfigs as per your suggestion. 
I have also addressed all your other comments - thanks for the quick review!

File: samples/modules/lvgl/accelerometer_chart/CMakeLists.txt [hunk 8]
```diff
@@ -0,0 +1,9 @@
+# SPDX-License-Identifier: Apache-2.0
+
+cmake_minimum_required(VERSION 3.20.0)
+
+find_package(Zephyr REQUIRED HINTS $ENV{ZEPHYR_BASE})
+project(acceleration_chart)
```
Comment (faxe1008) [inline samples/modules/lvgl/accelerometer_chart/CMakeLists.txt:6]: Nit: Maybe rename this also.

**Review by faxe1008 — DISMISSED** [review]
LGTM

File: samples/modules/lvgl/accelerometer_chart/CMakeLists.txt (hunk 8 above)
Comment (kartben) [reply samples/modules/lvgl/accelerometer_chart/CMakeLists.txt]: well spotted, thanks! also renamed the test case

File: samples/modules/lvgl/accelerometer_chart/src/main.c [hunk 9]
```diff
@@ -0,0 +1,97 @@
...
+static void sensor_timer_cb(lv_timer_t *timer)
+{
+	struct sensor_value accel[3];
+	int rc = sensor_sample_fetch(sensor);
+
+	if (rc == 0) {
+		rc = sensor_channel_get(sensor, SENSOR_CHAN_ACCEL_XYZ, accel);
+	}
+	if (rc < 0) {
+		LOG_ERR("ERROR: Update failed: %d\n", rc);
```
Comment (fabiobaltieri) [inline samples/modules/lvgl/accelerometer_chart/src/main.c:39]: use `return;` here and keep the normal execution path with no indentation, makes it easier to follow the indended code path.

File: samples/modules/lvgl/accelerometer_chart/src/main.c [hunk 10]
```diff
@@ -0,0 +1,97 @@
...
+	} else {
+		lv_chart_set_next_value(chart1, ser_x, sensor_value_to_double(&accel[0]));
+		lv_chart_set_next_value(chart1, ser_y, sensor_value_to_double(&accel[1]));
+		lv_chart_set_next_value(chart1, ser_z, sensor_value_to_double(&accel[2]));
+	}
+}
+
+void create_accelerometer_chart(lv_obj_t *parent)
```
Comment (fabiobaltieri) [inline samples/modules/lvgl/accelerometer_chart/src/main.c:47]: `static`

File: samples/modules/lvgl/accelerometer_chart/src/main.c [hunk 11]
```diff
@@ -0,0 +1,97 @@
...
+
+	/* Do not display point markers on the data */
+	lv_obj_set_style_size(chart1, 0, LV_PART_INDICATOR);
+}
+
+int main(void)
+{
+	const struct device *display_dev;
+
+	display_dev = DEVICE_DT_GET(DT_CHOSEN(zephyr_display));
```
Comment (fabiobaltieri) [inline samples/modules/lvgl/accelerometer_chart/src/main.c:73]: nit: use the same style to get the node for both display and sensor, either use an intermediate `_NODE` define for both or neither, just for coherency.

File: samples/modules/lvgl/accelerometer_chart/src/main.c (hunk 9 above)
Comment (kartben) [reply samples/modules/lvgl/accelerometer_chart/src/main.c:38]: +1

File: samples/modules/lvgl/accelerometer_chart/src/main.c (hunk 11 above)
Comment (kartben) [reply samples/modules/lvgl/accelerometer_chart/src/main.c:72]: ya this was actually bothering me as well :D

**Review by fabiobaltieri — DISMISSED** [review]

**Review by faxe1008 — DISMISSED** [review]

**Comment by kartben** [comment]
@fabiobaltieri @faxe1008  sorry for dismissing your reviews. Just pushed a small change to increase the main stack size to 4K, to be on the safe side. The sample didn't run out of the box on M5 Stack Core2, so I would suspect other ESP32, and possibly more, would be in the same situation. This is the same stack size as for the demos (and hopefully should be more than enough now).
Thanks!

**Review by faxe1008 — APPROVED** [review]

**Review by fabiobaltieri — APPROVED** [review]

**Pull request merged**